lessons/r/intro/lesson.Rmd

```diff
@@ -0,0 +1,555 @@
+2 > 3
+7 < 9
+2 <= 4 # the equals sign comes second
+2 >= 4 # equals sign still comes second!
+3 %in% c(2, 3, 4) # check for membership
+is.numeric(3) # check for an object type
+is.character(3)
+```
+
+Back to `dplyr`!
+
+## Chaining operations - meet the pipe
+
+`select` and `filter` are useful and all, but what happens
+when you want to subset your data frame by columns AND a condition?
+
+We could try nesting the functions:
+
+```{r}
+filter(select(airquality, Month, Day), Month > 6)
+```
+
+But this is unwieldy to both read and write.
+
+What about a temporary object?
+
+```{r}
+temp <- select(airquality, Month, Day)
+filter(temp, Month > 6)
+```
+
+This works, but we don't want to keep making new temporary objects
+for every single operation. Not only is it tedious and hard to keep
+track of as our analyses proceed, it can also quickly drain your
+computer's memory.
+
+So what's the most efficient way to make this work? Well, `dplyr`
+features a special operation just for this purpose, known as
+the 'pipe', or `%>%`. Since that can be a bit annoying to type
+out, RStudio features a handy shortcut - Ctrl/Cmd + Shift + M. 
+
+The pipe will take whatever's on the left and feed it as input
+to whatever is on the right. In other words, the following two 
+expressions are equivalent:
+
+```{r}
+select(airquality, Month, Day)
+airquality %>% select(., Month, Day)
+```
+
+If the dot is at the start of the function, we can leave it out
+as shorthand:
+
+```{r}
+airquality %>% select(Month, Day)
+```
+
+We can use the pipe to combine `select` and `filter`:
+
+```{r}
+airquality %>% 
+  select(Month, Day) %>% 
+  filter(Month > 6)
+```
+
+And this is the magic of `dplyr`! Using the pipe operator, we can
+chain successive functions and run our data frame through a series
+of useful operations.
+
+Let's cover the remaining `dplyr` verbs:
+
+## `arrange` - sort by column
+
+`arrange` returns a data frame sorted in ascending order by the specified column.
+
+```{r}
+airquality %>% 
+  arrange(Temp)
+```
+
+Sorting by multiple columns sorts by each in order in a 'nested' fashion:
+
+```{r}
+airquality %>% 
+  arrange(Temp, Wind)
+```
+
+Finally, we can sort in descending order using the `desc()` helper function:
+
+```{r}
+airquality %>% 
+  arrange(desc(Month))
+```
+
+## `mutate` - apply windowed functions
+
+So far, we've explored functions that allow us to subset data frames and modify them
+as they are. Where `dplyr` arguably comes to life is its ability to actually perform
+vectorized operations across entire columns of a given data frame via the `mutate` function.
+
+The `Temp` column in our data frame contains temperature values in Fahrenheit. How might
+we convert these values to Celsius instead?
+
+Recall that an approximate conversion between the two is as follows:
+
+$$ C = (F - 32) / 1.8 $$
+Let's perform this conversion with `mutate`:
+
+```{r}
+airquality %>% 
+  mutate(temp_celsius = (Temp - 32) / 1.8)
```

Comment:
If we wanted to get fancy, we could assign the conversion to a function (e.g. `far_to_celsius`) and use the custom function in `mutate`. Would highlight a few concepts at once.